### DISCUSSION TITLE
Node.js Client API Problems

### DISCUSSION THREAD

**Discussion by mkastner** [duckdb/duckdb#7067] [General]
I don't know who maintains the node.js client API so I keep my remarks rather general.

My intention was to use duckdb in a node.js for a node.js microservice, which aggregates some data. But after trying for a few days I had to give up, since I wasn't able to integrate the node.js client API into my build process.

One of the reasons were continuing errors regarding two AWS-dependencies and a nock dependency in development.

Since I am not interested in AWS and my integration tests don't require nock I don't know why that stuff is needed.

The other thing is using node-gyp for binary builds. node-gyp is a notorious source build failure and is buggy as hell. I haven't really checked for its purpose, but it failed, at whatever job it was intended for.

As much as I'd have loved to use DuckDB for that project, I wasn't able to put it to use.

Maybe that's just me, then please, ignore my outbreak of frustration.

**Comment by Mause** [comment]
Hi @mkastner 

Thanks for the feedback, it's appreciated.

Your frustrations regarding the dependencies that @mapbox/node-pre-gyp requires are unfortunately shared by many other people, and the library itself is barely maintained, so issues have been opened and have remained open for quite some time, see https://github.com/mapbox/node-pre-gyp/issues/661. I've looked into other tools in the past, but they unfortunately require a great deal of change in how we build and distribute the library.

Regarding node-gyp, it bills itself as the "Node.js native addon build tool", and is the defacto way to compile C/C++ into a native library extension that NodeJS can load. DuckDB itself is written in C++, so its use is not optional.

If you have the time and energy to do so, we'd appreciate you posting the compilation errors you received and on which platform and architecture, as ideally node-pre-gyp would avoid needing to compile the library from scratch

Many thanks

**Reply by mkastner** [reply]
Thank you very much for your kind feedback. I more than appreciate your explanations especially since my original post might have come across somewhat with a bad temper.

It might not be obvious to Linux/Macos or even Windows users. I run most of my servers on FreeBSD and node-gyp is a nightmare on that system and therefor I try to avoid dependencies on requiring node-gyp. I know it works fine on most debian and debian based Linuxes. But not on FreeBSD and I've searched the hell out of google for solutions.

In the case of duckdb, I've already run into problems on my MacOS-development machine, so I expect worse stuff to happen on FreeBSD and BSDs in general for that matter.

That being said, I'll keep an eye on DuckDB. It's really awesome. Thanks to all participants for the great work.
